src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp

```diff
@@ -0,0 +1,610 @@
+                bool hasUnwanted = hasUnwantedConnections(b1ForwardCandidatePrimerSequence);
+                if (!hasUnwanted) {
+                    // If there are no unwanted connections - we are found b1 forward candidate primer
+                    // Now we need to check b1 reverse primer
+                    // b1 forwards and b1 reverse shouldn't have unwanted hetero-dimers - if they do we need to continue searching of b1 forward
+                    findB1ReversePrimer(b1ForwardCandidatePrimerSequence);
+
+                    //If b1 reverse wasn't found - continue searching
+                    if (b1Reverse.isEmpty()) {
+                        taskLog.details(tr("The \"B1 Forward\" primer with the region \"%1\" doesn't fit because there are no corresponding \"B1 Reverse\" primers").arg(QString(b1ForwardCandidatePrimerRegionString)));
+                        updatePrimerRegion(b1ForwardCandidatePrimerEnd, b1ForwardPrimerLength);
+                        continue;
+                    }
+
+                    //If it was found - we found b1 forward
+                    b1Forward = b1ForwardCandidatePrimerRegion;
+
+                    //The first time we found the pair of B1 primers - they are also A primers
+                    if (aWasNotFoundYet) {
+                        taskLog.details(tr("The \"A Forward\" and the \"A Reverse\" primers have been found"));
+                        aForward = b1Forward;
+                        aReverse = b1Reverse;
+                        aWasNotFoundYet = false;
+                    }
+
+                    //Now we need to find B2 and B3
+                    findSecondaryForwardReversePrimers(SecondaryPrimer::B2);
+                    findSecondaryForwardReversePrimers(SecondaryPrimer::B3);
+
+                    // We need to find at leas one pair - B2 or B3 - for B1 to be valid
+                    // If we didn't - continue searching
+                    if (b2Forward.isEmpty() && b3Forward.isEmpty()) {
+                        taskLog.details(tr("B2 and B3 primer pairs haven't been found, search again"));
+                        b1Forward = U2Region();
+                        b1Reverse = U2Region();
+                        updatePrimerRegion(b1ForwardCandidatePrimerEnd, b1ForwardPrimerLength);
+                        continue;
+                    }
+
+                    break;
+                } else {
+                    taskLog.details(tr("The candidate primer region \"%1\" contains unwanted connections").arg(QString(b1ForwardCandidatePrimerRegionString)));
+                    //If they are unwanted connections - update primer region and check all parameters again
+                    updatePrimerRegion(b1ForwardCandidatePrimerEnd, b1ForwardPrimerLength);
+                    continue;
+                }
+            }
+        }
+        if (!b1Forward.isEmpty()) {
+            break;
+        }
+    }
+    generateUserPrimersReports();
+    stateInfo.setProgress(100);
```

Comment:
Not needed: set by the scheduler automatically

Reply:
OK